### PR TITLE
feat(search): boost current product section in Algolia results

### PR DESCRIPTION
**Pull request by jennifer-shehane**
## Summary

Two related changes to the Algolia docs search:

1. **Section-aware result boosting.** Docs search now soft-boosts results from the product section the reader is currently in (App, API, Cloud, UI Coverage, or Accessibility) to the top of the results list, while keeping every other result visible below. Implemented in the swizzled `SearchBar`'s `transformItems` hook — no crawler-config change and no reindex required.
2. **Unit tests for the Algolia search + scraper scripts**, with light refactors to make that logic testable in isolation.

## Why

All five products live in a single flat Algolia index (`cypress_docs`), so a query like `configuration` or `timeout` returns results mixed across every product regardless of where the reader is browsing. Someone reading the API reference searching `timeout` most likely wants the API result, but it competes flatly with App/Cloud/etc. hits.

The natural fix is Algolia's server-side `optionalFilters` boost, but probing the live index showed it returns `{"message":"optionalFilters is not supported on this plan"}`. A hard `facetFilters` scope is supported but would *hide* other sections, a poor default for cross-product discovery. So this achieves the same "prefer my current section" behavior client-side: a stable partition on `hierarchy.lvl0` (whose exact index values — `App`, `API`, `Cloud`, `UI Coverage`, `Accessibility` — were confirmed against the live index) that reorders without dropping any result.

Closes #

## What's tested

`npm run test:search` — 49 vitest cases, also wired into CI as a **"Unit Tests (Search/Algolia)"** job:

- **`searchRanking`** (extracted from `SearchBar/index.js` into `searchRanking.js`): path→section mapping, stable/lossless section boosting, facet-filter merging.
- **`algolia-client.mjs`**: constructor validation, pure `selectIndexEntries`, and `getEntriesForIndex` driven by an injected fake fetch (no network). Also fixes a latent bug — the index-not-found error interpolated the `undefined` lookup result instead of the requested index name.
- **`scrape-and-compare-algolia-index.mjs`**: `parseScraperHits` and a newly-extracted pure `evaluateScrape()` covering every verdict (ok / upload-mismatch / index-drop / large-increase), precedence, the ratio-scaled tolerance, and boundary conditions. `main()` is now guarded so importing the module for tests runs no I/O.

## Notes for reviewers

- **Boost is lossless & stable:** it only moves current-section hits ahead of the rest; relevance order within each group is preserved and the result count never changes. Outside the five sections (home, `/search`) the section is `null` and results pass through untouched.
- **Why a ref:** `transformItems` is frozen at first render (`useRef(fn).current`), so the current section is written to `currentSectionRef` each render (mirroring the existing `queryIDRef` pattern) to stay live across client-side navigation.
- **Refactors are behavior-preserving** aside from the intentional error-message bug fix; the scraper's runtime path is unchanged (verified `main()` still runs on direct invocation, and a production `npm run build` passes).
- **Ceiling:** the boost reorders the result window DocSearch fetches, so if the relevant page is never returned it can't surface it — the limitation of doing this client-side instead of via the plan-gated `optionalFilters`.
- **Out of scope (need Algolia dashboard access):** enabling Dynamic Re-Ranking on the click-analytics already being collected, and Query Rules / synonym updates driven by no-result queries.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01YDBrhUnf3j6bj84VReDBgZ

---

> [!NOTE]
> **Medium Risk**
> Search UX and analytics positions change for all DocSearch users; synonym/config changes affect future scrapes. Core logic is well-tested and scraper runtime behavior is largely preserved aside from the intentional error-message fix.
> 
> **Overview**
> Docs search now **soft-boosts hits from the reader’s current product section** (App, API, Cloud, UI Coverage, Accessibility) to the top of the modal list without hiding other sections. The swizzled `SearchBar` derives section from `useLocation`, runs `boostCurrentSection` and `assignDisplayPositions` in `transformItems`, and reports **`__displayPosition`** to Algolia click analytics so positions match what users saw after reordering. Ranking helpers live in new `searchRanking.js`.
> 
> **CI and tests:** `npm run test:search` runs ~49 Vitest cases; CircleCI adds a **Unit Tests (Search/Algolia)** job after `build`. `algolia-client.mjs` gains injectable fetch and exported `selectIndexEntries` (fixes index-not-found error text). The scrape validator exports `parseScraperHits` / `evaluateScrape` and only runs `main()` on direct invocation.
> 
> **Index tuning (reindex-related):** `config.json` adds many Algolia **synonyms**; new `query-rules.json` documents dashboard query-cleanup rules (not wired in-repo).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8b0b9517308ddb955d5edd532deeec875f77d728. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: scripts/search/scrape-and-compare-algolia-index.mjs [hunk 1]
```diff
@@ -95,52 +190,41 @@ const main = async () => {
...
   process.exit(0)
 }
 
-main()
+// Only run the scraper when invoked directly (e.g. from CI), not when this
+// module is imported by tests.
+const invokedDirectly =
+  process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]
+if (invokedDirectly) {
+  main()
```
[the remark this answers is not in the document]
Comment (jennifer-shehane) [reply scripts/search/scrape-and-compare-algolia-index.mjs:233]: Fixed in 5cf68fd. Both sides of the guard are now resolved to an absolute path before comparing:

```js
const invokedDirectly =
  process.argv[1] &&
  path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)
```

`path.resolve` is a no-op on an already-absolute path and resolves a relative one against cwd, so the CI invocation (`node ./scripts/search/scrape-and-compare-algolia-index.mjs` from the repo root) now matches. Verified: a relative invocation runs `main()` (fails on the missing key, as expected), while importing the module still does not.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/theme/SearchBar/index.js [hunk 2]
```diff
@@ -193,20 +236,20 @@ function DocSearch({ contextualSearch, externalUrlRegex, ...props }) {
           // Attach queryID to each hit for click tracking (if available)
           queryID: queryIDRef.current || item.queryID,
         }))
-    
-    return transformedItems
+
+    // Prefer results from the product section the reader is currently in.
+    return boostCurrentSection(transformedItems, currentSectionRef.current)
```
[the remark this answers is not in the document]
Comment (jennifer-shehane) [reply src/theme/SearchBar/index.js]: Good catch — fixed in 5cf68fd. After `boostCurrentSection` reorders the hits, `transformItems` now stamps each hit with its final displayed 1-based position via `assignDisplayPositions`, and the click handler reads that first:

```js
positions: [
  hit.__displayPosition ||
    hit.__position ||
    hit.__autocomplete_id ||
    hit.position ||
    1,
],
```

So a hit Algolia ranked 3rd that gets boosted to the top now reports position 1 to `clickedObjectIDsAfterSearch`, matching what the user actually saw. Covered by new unit tests in `searchRanking.test.js`.

---
_Generated by [Claude Code](https://claude.ai/code)_
